lib/Runtime/Provisioner/Provisioner.cpp

```diff
@@ -58,12 +58,13 @@ llvm::Error Provisioner::provision(DAGListTy &networks, Module &module) {
 
   for (auto &network : networks) {
     for (auto &node : network.nodes) {
-      auto it = logicalDevices.find(node->logicalDevices[0]);
-      if (it != logicalDevices.end()) {
-        it->second.push_back(node.get());
-      } else {
-        logicalDevices.emplace(node->logicalDevices[0],
-                               std::vector<DAGNode *>{node.get()});
+      for (auto logical : node->logicalDevices) {
+        auto it = logicalDevices.find(logical);
+        if (it != logicalDevices.end()) {
+          it->second.push_back(node.get());
+        } else {
+          logicalDevices.emplace(logical, std::vector<DAGNode *>{node.get()});
```

Comment:
Are we creating an std::vector here? Why? Why can't we allocate a local static array and pass some ArrayRef or Begin-End pair? std::vector will cause a memory allocation.

Reply:
The provisioner gives each DagNode a logical device id and this is grouping the DagNodes by those ids. Because these logical devices are created dynamically per graph it would be difficult to have a static allocation for them

Reply:
ah, sorry. I missed the fact that logicalDevices is a vector of vectors.